drivers/espi/espi_emul.c

```diff
@@ -19,168 +19,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+	 */
+	data->kbc8042_dbbin = 0;
+	data->kbc8042_status &= ~(KBC8042_STATUS_A2 | KBC8042_STATUS_IBF);
+}
 
-static int emul_espi_trigger_event(const struct device *dev,
-				   struct espi_event *evt)
+static void espi_emul_send_obe_irq(const struct device *dev)
 {
 	struct espi_emul_data *data = dev->data;
+	struct espi_event ev;
 
-	if (((evt->evt_type & ESPI_BUS_EVENT_VWIRE_RECEIVED) &&
-	     !(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) ||
-	    ((evt->evt_type & ESPI_BUS_EVENT_OOB_RECEIVED) &&
-	     !(data->cfg.channel_caps & ESPI_CHANNEL_OOB)) ||
-	    ((evt->evt_type & ESPI_BUS_PERIPHERAL_NOTIFICATION)
-	     && !(data->cfg.channel_caps & ESPI_CHANNEL_PERIPHERAL))) {
-		return -EIO;
+	if (!data->kbc8042_irq_enabled || !data->kbc8042_obe_irq_pending) {
+		return;
 	}
 
-	espi_send_callbacks(&data->callbacks, dev, *evt);
+	data->kbc8042_obe_irq_pending = 0;
+
+	ev.evt_type = ESPI_BUS_PERIPHERAL_NOTIFICATION;
+	ev.evt_details = ESPI_PERIPHERAL_8042_KBC;
+
+	struct espi_evt_data_kbc *kbc_evt =
+				(struct espi_evt_data_kbc *)&ev.evt_data;
+	kbc_evt->evt = HOST_KBC_EVT_OBE;
+	kbc_evt->data = 0;
+	kbc_evt->type = 0;
+
+	espi_send_callbacks(&data->callbacks, dev, ev);
+}
+#endif
+
+/** API **/
+
+int espi_emul_config(const struct device *dev, struct espi_cfg *cfg)
+{
+	struct espi_emul_data *data = dev->data;
+
+	data->channels_enabled = cfg->channel_caps;
 
 	return 0;
 }
 
-static bool espi_emul_get_channel_status(const struct device *dev, enum espi_channel ch)
+bool espi_emul_get_channel_status(const struct device *dev,
+	enum espi_channel ch)
 {
 	struct espi_emul_data *data = dev->data;
 
-	return (data->cfg.channel_caps & ch);
+	if (data->channels_enabled & ch) {
+		return 0;
+	}
+
+	return -EIO;
 }
 
-static int espi_emul_read_lpc_request(const struct device *dev,
-				      enum lpc_peripheral_opcode op,
-				      uint32_t *data)
+/* Logical Channel 0 APIs */
+int espi_emul_read_request(const struct device *dev,
+	struct espi_request_packet *req)
 {
-	const struct emul_espi_device_api *api;
-	struct espi_emul *emul;
-	struct espi_emul_data *emul_data = dev->data;
+	return -EIO;
+}
 
-	if (!(emul_data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
-		LOG_ERR("bad channel vwire");
-		return -EINVAL;
+int espi_emul_write_request(const struct device *dev,
+	struct espi_request_packet *req)
+{
+	return -EIO;
+}
+
+int espi_emul_lpc_read_request(const struct device *dev,
+	enum lpc_peripheral_opcode op, uint32_t *value)
+{
+	struct espi_emul_data *data = dev->data;
+	struct espi_emul *emul = data->emul;
+
+	if (!(data->channels_enabled & ESPI_CHANNEL_PERIPHERAL)) {
+		return -EIO;
 	}
 
-	emul = espi_emul_find(dev, EMUL_ESPI_HOST_CHIPSEL);
-	if (!emul) {
-		LOG_ERR("espi_emul not found");
-		return -ENOTSUP;
+	switch (op) {
+#ifdef CONFIG_ESPI_EMUL_KBC8042
+	case E8042_OBF_HAS_CHAR:
+		*value = !!(data->kbc8042_status & KBC8042_STATUS_OBF);
+		break;
+	case E8042_IBF_HAS_CHAR:
+		*value = !!(data->kbc8042_status & KBC8042_STATUS_IBF);
```

Comment:
Hm, are you sure that it will raise the error? I've done a little test with gcc 9.4.0 and doesn't fail to compile:
```
        uint8_t x = 5;
        int y = 9;
        uint32_t v;
        uint32_t* ptr = &v;

        *ptr = !!(x & y);
        printf("V: %d", v);
```

Also it looks like the "!!(x&y)" is used often in the zephyr drivers, even in ITE and NPCX espi drivers.